fix(ci): add branch restriction and label override for artifact fallback

aliersh opened the pull request:
This PR adds branch-based conditional logic to restrict artifact fallback to PR builds only, and implements a label-based override mechanism for emergency cases.

### Reason

The fallback mechanism (which uses latest develop artifacts for speed) needs two safety controls:

1. **Branch restriction**: Develop merges must always build with fresh artifacts to avoid issues with snapshots/checks jobs that are sensitive to stale artifacts
2. **Label override**: PRs need an escape hatch to disable fallback when stale artifacts cause test failures or unexpected behavior

These controls maintain CI health on develop while preserving PR optimization benefits, with a safety valve for edge cases.

### Changes

- Add branch-based conditional logic to check `CIRCLE_BRANCH != develop` before enabling `--fallback-to-latest`
- Add `force-use-fresh-artifacts` PR label check to allow manual override of fallback (requires manual CI rerun after adding label)
- Extract `pull-artifacts-conditional` reusable CircleCI command to avoid code duplication
- Create `use-latest-fallback.sh` bash script with branch and label checking logic
- Update `contracts-bedrock-tests`, `contracts-bedrock-coverage`, and `contracts-bedrock-tests-upgrade` jobs to use new command

aliersh committed:
fix(ci): restrict artifact latest fallback to PRs only

- add branch check to only use --fallback-to-latest on non-develop branches
- ensure develop merges build with fresh artifacts

aliersh committed:
feat(ci): add force-use-fresh-artifacts label override for artifact fallback

smartcontracts commented on .circleci/config.yml:
```diff
@@ -340,6 +340,35 @@ commands:
             - ~/go/pkg/mod
           key: go-<<parameters.namespace>>-<<parameters.version>>-<<parameters.module>>-{{ checksum "<<parameters.module>>/go.mod" }}-{{ checksum "<<parameters.module>>/go.sum" }}
 
+  pull-artifacts-conditional:
```
LGTM but would be ideal for @raffaele-oplabs to review

raffaele-oplabs commented on .circleci/config.yml:
```diff
@@ -340,6 +340,35 @@ commands:
+              # Check if PR has force-use-fresh-artifacts label (override fallback)
+              if [ -n "${CIRCLE_PULL_REQUEST}" ]; then
+                export GH_TOKEN=${MISE_GITHUB_TOKEN}
+                if gh pr view ${CIRCLE_PULL_REQUEST} --json labels | grep -q "force-use-fresh-artifacts"; then
```
The CI fail here: https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/108033/workflows/918d70c8-f165-4594-a262-671e0da3e07e/jobs/4113043?invite=true#step-116-0_41

I would suggest using this approach with curl: https://github.com/ethereum-optimism/optimism/pull/18330/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R236

Where possible to have simple curl against an API doing the same job I would go with that to avoid having to maintain/download extra dependencies, with their versions.

aliersh replied:
@raffaele-oplabs fixed in [f62605d](https://github.com/ethereum-optimism/optimism/pull/18330/commits/f62605dc4065cea47f49e3f9ef7089b7c6677eaf)

raffaele-oplabs reviewed (changes requested):
See comment in the review

aliersh committed:
fix(ci): replace gh cli with curl for label check

aliersh commented:
I had to debug the logic because the API call didn't fetch the label's name. Fixed in [2ec3c05](https://github.com/ethereum-optimism/optimism/pull/18330/commits/2ec3c05aa6030eec98d133a6a66fc78825238926)

maurelian reviewed (changes requested):
We want to [avoid writing logic in the CI config file](https://www.notion.so/oplabs/P-PS-Distributed-CI-Ownership-278f153ee16280bda58ff28a76a86946?source=copy_link#278f153ee16280f8a29cdcfecaf54245) and should at a minimum put it into a just command.

In this case I actually think there should be bash script, which is callable by a just command, which is run in the CI config file.

aliersh committed:
refactor(ci): extract artifact fallback logic to bash script

- move conditional logic from circleci config to use-latest-fallback.sh

aliersh commented:
> We want to [avoid writing logic in the CI config file](https://www.notion.so/oplabs/P-PS-Distributed-CI-Ownership-278f153ee16280bda58ff28a76a86946?source=copy_link#278f153ee16280f8a29cdcfecaf54245) and should at a minimum put it into a just command.
> 
> In this case I actually think there should be bash script, which is callable by a just command, which is run in the CI config file.

@maurelian fixed in [66ed564](https://github.com/ethereum-optimism/optimism/pull/18330/commits/66ed5640dd7388cebb2e0ebddad1a52b7243e30b)

aliersh committed:
fix(ci): add security flags to curl in use-latest-fallback.sh

aliersh commented on packages/contracts-bedrock/scripts/ops/use-latest-fallback.sh:
```diff
@@ -0,0 +1,36 @@
+
+    # Query GitHub API for PR details
+    PR_DATA=$(curl -s -H "Authorization: token ${MISE_GITHUB_TOKEN}" \
+      "https://api.github.com/repos/ethereum-optimism/optimism/pulls/${PR_NUMBER}")
```
fixed in [e598095](https://github.com/ethereum-optimism/optimism/pull/18330/commits/e598095253e98f0983733374990472dca0ac36fa)

maurelian commented on packages/contracts-bedrock/scripts/ops/use-latest-fallback.sh:
```diff
@@ -0,0 +1,36 @@
+    PR_DATA=$(curl -sS --fail --connect-timeout 10 --max-time 30 -H "Authorization: token ${MISE_GITHUB_TOKEN}" \
+      "https://api.github.com/repos/ethereum-optimism/optimism/pulls/${PR_NUMBER}")
+
+    if echo "$PR_DATA" | grep -q '"name"[[:space:]]*:[[:space:]]*"force-use-fresh-artifacts"'; then
```
grepping through the API response feels pretty brittle to me. I see that you [previously](https://github.com/ethereum-optimism/optimism/pull/18330/commits/2ec3c05aa6030eec98d133a6a66fc78825238926) were using the `gh` cli which is nicer, but I guess we don't assume it is available. IMO `jq` would be a more robust approach vs `grep`. This is the suggestion from copilot:

```
GitHub Copilot Chat Assistant — Use jq to parse the labels instead of grep. Replace the curl + grep check with a jq-based check that looks for a label named "force-use-fresh-artifacts".

Suggested replacement (concise):

PR_DATA=$(curl -sS --fail --connect-timeout 10 --max-time 30 \
  -H "Authorization: token ${MISE_GITHUB_TOKEN}" \
  "https://api.github.com/repos/ethereum-optimism/optimism/pulls/${PR_NUMBER}")

if echo "$PR_DATA" | jq -e 'any(.labels[]; .name == "force-use-fresh-artifacts")' >/dev/null; then
  echo "Force use fresh artifacts label detected, skipping fallback"
  USE_FALLBACK=false
fi

Notes:
- jq -e returns exit code 0 when the expression evaluates to true, so the shell if works as expected.
- any(.labels[]; .name == "...") checks whether any label object has that name.
- You can call the smaller endpoint that returns only labels if you prefer: "https://api.github.com/repos/ethereum-optimism/optimism/issues/${PR_NUMBER}/labels" and then run jq -e 'any(.[]; .name == "force-use-fresh-artifacts")'.
- Ensure jq is installed in your CI image and MISE_GITHUB_TOKEN is set.
```

aliersh replied:
@maurelian Good point! Yes, I did it first with `gh` cli but I changed it because it wasn't available. Switch to `jq` seems pretty reasonable

aliersh replied:
fixed in [2cef659](https://github.com/ethereum-optimism/optimism/pull/18330/commits/2cef659a779f93a6507af939ec317ff09de8fa79)

aliersh committed:
fix(ci): use jq instead of grep for json parsing

aliersh committed:
fix(ci): add error handling for github API failures

- wrap curl and jq calls in conditional to catch failures
- default to fallback behavior if API call fails (safe default)
- log warning message when API is unreachable

aliersh commented on packages/contracts-bedrock/scripts/ops/use-latest-fallback.sh:
```diff
@@ -0,0 +1,36 @@
+    # Query GitHub API for PR details
+    PR_DATA=$(curl -sS --fail --connect-timeout 10 --max-time 30 -H "Authorization: token ${MISE_GITHUB_TOKEN}" \
+      "https://api.github.com/repos/ethereum-optimism/optimism/pulls/${PR_NUMBER}")
+
```
fixed in [d094a23](https://github.com/ethereum-optimism/optimism/pull/18330/commits/d094a23dff099177474e52d676b45f289b695e2f)
